### ISSUE TITLE
DL4J: Add a SameDiffLoss class for DL4J loss functions based on SameDiff

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#8652]
Loss functions for DL4J are defined here: https://github.com/eclipse/deeplearning4j/tree/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl

Though these work fine, there is no easy way to add custom loss functions for DL4J using SameDiff, outside of implementing a whole SameDiffOutputLayer instance.
And if users don't use SameDiff, they have to work out the gradients and do implement backprop manually.

Instead, what we want here: an abstract class called SameDiffLoss, that implements ILossFunction. It should be defined here:
https://github.com/eclipse/deeplearning4j/tree/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions

The user (who would extend SameDiffLoss) should have to define one method in order to create a loss function:
```
defineLoss(SameDiff sameDiff, SDVariable layerInput, SDVariable labels)
```
The remaining methods from ILossFunction, such as computeScore, computeScoreArray, computeGradient, computeGradientAndScore, and name() should be implemented in SameDiffLoss.

Once the implementation is complete, let's add a couple of implementations using/extending SameDiffLoss (such as MSE), just in the tests, and gradient check them here:
https://github.com/eclipse/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-core/src/test/java/org/deeplearning4j/gradientcheck/LossFunctionGradientCheck.java

**Comment by rhoegg** [comment]
I'm kind of new to deeplearning4j, and I see that we have `SameDiffLoss` as of 1.0.0-M2.1  (thank you!)

I tried implementing huber loss like so. But I ran into trouble saving and loading `MultiLayerNetwork` when I did this.  I see this issue is still open, does the ability to save/load a model using a custom loss function belong as part of this issue, or is it a separate concern?

```java
OutputLayer.Builder builder = new OutputLayer.Builder(new SameDiffLoss() {
    @Override
    public SDVariable defineLoss(SameDiff sd, SDVariable layerInput, SDVariable labels) {
        return sd.loss.huberLoss(labels, layerInput, null, huberLossDelta);
    }
});
```

**UPDATE: this was a PEBKAC, future readers please see my next comment**

**Comment by rhoegg** [comment]
For my earlier issue with saving, all I needed to do was to make my `SameDiffLoss` subclass properly implement `java.io.Serializable` and it loaded fine.


Since `SameDiffLoss` seems to be working, can this issue be closed?

**Comment by agibsonccc** [comment]
Sounds great. Thanks! Let me know if you have any other issues.
